[Federation] Kubefed init verifies if control plane pods are up before returning success

irfanurrehman opened the pull request:
This PR updates the functionality as needed in issue https://github.com/kubernetes/kubernetes/issues/37841.

cc @kubernetes/sig-cluster-federation @nikhiljindal @madhusudancs @shashidharatd

Commit:
[Federation] Wait for control plane pods in kubefed init

Commit:
[Federation] Unit tests updated for wait for control plane pods in kubefed init

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/39230)
<!-- Reviewable:end -->

madhusudancs commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -230,6 +230,23 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
 
 	if !dryRun {
+		fmt.Fprintf(cmdOut, "Waiting for control plane to come up")
+		for !podRunning(hostClientset, serverName, initFlags.FederationSystemNamespace) {
```
Please use https://godoc.org/k8s.io/kubernetes/pkg/util/wait#Poll or https://godoc.org/k8s.io/kubernetes/pkg/util/wait#PollImmediate instead.

madhusudancs reviewed (commented):
This might be useful, but isn't sufficient. What we should really do is component health checks, in the same vein as `ComponentStatus`es reported by Kubernetes components. We are not there yet, so what we should do is pod status check you are doing here plus also verify that an API endpoint responds. See here - https://github.com/kubernetes/kubernetes/pull/34484/files#diff-776638e5b09cfdf82b07554f70d6b558R430

Commit:
[Federation] Review comment fixes for wait for control plane pods in kubefed init

irfanurrehman commented:
@madhusudancs thanks for the comments, have updated accordingly, re-review needed!!

madhusudancs commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -229,7 +230,25 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
 		return err
 	}
 
+	//to talk to federation ctrl plane
```
Let's try to avoid arbitrary acronyms and vowel-omissions in the comments in the future. Comments are there to explain what is happening in the code. Let's not make it difficult to even read them.

irfanurrehman replied:
sure madhu; will update with mode elaborate comment, and will take care of this in future also.

madhusudancs commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -229,7 +230,25 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
 	}
 
+	//to talk to federation ctrl plane
+	fedFactory := config.HostFactory(initFlags.Name, initFlags.Kubeconfig)
```
Why build it here and not inside the `if`-block, right before the call to `waitSrvHealthy()`? I think it is Ok to just pass the context name and kubeconfig you are passing to `HostFactory()` to `waitSrvHealthy()` instead and build the client object there. 

Also, I am somewhat worried about doing this. `HostFactory()` method was written keeping interactions with host cluster in mind. It wasn't for federation API server. Also, the discovery interface for the command factory builds the objects for Kubernetes APIs and not federation APIs. All of these might work today because of the particular way things are hooked up, but might break down when some of these assumptions are violated.

In order to avoid this problem in the future and also to do the right thing, directly build the federation Clientset in `k8s.io/kubernetes/federation/client/clientset_generated/federation_clientset`. You can then call `fedClientset.Discovery()` to get the federation discovery client.

irfanurrehman replied:
> Why build it here and not inside the if-block, right before the call to waitSrvHealthy()? I think it is Ok to just pass the context name and kubeconfig you are passing to HostFactory() to waitSrvHealthy() instead and build the client object there.

Will update accordingly!

irfanurrehman replied:
> Also, I am somewhat worried about doing this. HostFactory() method was written keeping interactions with host cluster in mind. It wasn't for federation API server. Also, the discovery interface for the command factory builds the objects for Kubernetes APIs and not federation APIs. All of these might work today because of the particular way things are hooked up, but might break down when some of these assumptions are violated.
> 
> In order to avoid this problem in the future and also to do the right thing, directly build the federation Clientset in k8s.io/kubernetes/federation/client/clientset_generated/federation_clientset. You can then call fedClientset.Discovery() to get the federation discovery client.

I have a slight confusion about this. Please help me understand better.
Actually when I implemented it first, I did something similar what are asking for. I did extend the AdminConfig interface to add another method getconfig, which built the config on similar lines as is currently done to get hostfactory, and then got a separate client for fed control plane from that config. I however removed that in favor of simplicity of the current one.
Also on futher checking, the HostFactory() impl gets the new factory from package kubectl/cmd which I guess "should" work with either federation API server or kube API server, for that particular version alike.

I would any ways keep the impl as suggested by you also ready, and update this commit or not, based on your suggestion.

madhusudancs commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -582,6 +601,42 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
 	return clientset.Extensions().Deployments(namespace).Create(dep)
 }
 
+func waitForPods(clientset *client.Clientset, name, nameSpace string) error {
```
Why not make `waitForPods()` take a slice of names `waitForPods(..., name []string, ...)` and check for both the pods together instead of having to call them separately? Pluralized name for the function is already suggesting that.

Also, convention is to call the variable `namespace` - no capitalization for `S`.

irfanurrehman replied:
will update accordingly!!

madhusudancs commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -582,6 +601,42 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
+			return false, nil
+		}
+		for _, pod := range podList.Items {
+			if strings.Contains(pod.Name, name) && pod.Status.Phase == "Running" {
```
I have been thinking quite a bit about this. I think the right thing to do here is to get the deployment status, then get the pod by using its selector and then check the pod's phase. It's Ok to do that in a follow up PR. But at least do a `strings.HasPrefix()` check instead of a `strings.Contains()` check here.

irfanurrehman replied:
will update to use HasPrefix now and update rest of the checking method in a separate PR.

madhusudancs commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -582,6 +601,42 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
+		body, err := fedDiscoveryClient.RESTClient().Get().AbsPath("/healthz").Do().Raw()
+		if err != nil {
+			return false, nil
+		} else {
```
Drop the `else` here. It is not required. Logic will still be the same.

irfanurrehman replied:
sure!

madhusudancs commented on federation/pkg/kubefed/init/init_test.go:
```diff
@@ -168,7 +168,7 @@ func TestInitFederation(t *testing.T) {
 			}
 
-			if got := buf.String(); got != want {
+			if got := buf.String(); !strings.Contains(got, want) {
```
Why this change?

irfanurrehman replied:
Earlier kubefed init will either return a success statement exactly as implemented in printSuccess() or some failure message.
I did find some sense in adding additional messages before it (eg - Waiting for control plane to come up) , else it seemed like hung, if the wait takes long.
I did contemplate adding a "." every second or two to reflect that something is still happening, but did not include it in the final PR somehow.
The check in that case cannot be strictly only on the specific message.
What do u suggest?, should I add that back. or put a strict check on the message on the unit test.

irfanurrehman replied:
have updated accordingly, will raise another PR for this change.

madhusudancs commented on federation/pkg/kubefed/init/init_test.go:
```diff
@@ -694,6 +694,38 @@ func fakeInitHostFactory(federationName, namespaceName, ip, dnsZoneName, image,
+		},
+	}
+
+	ctrlMgrPod := v1.Pod{
```
`cm` is the convention used for controller manager.

irfanurrehman replied:
will update!

madhusudancs reviewed (changes requested):
I have left a few comments. Let me know once you have addressed them.

irfanurrehman commented:
@madhusudancs , I have updated the PR as per you review comments and suggestions.
As of now, I have added another commit to make it easier for you track changes.
One you are ok and say so, I will squash the review fixes commits into 1.
thanks!

madhusudancs commented:
@k8s-bot ok to test

madhusudancs commented on federation/pkg/kubefed/util/util.go:
```diff
@@ -64,17 +68,30 @@ func (a *adminConfig) PathOptions() *clientcmd.PathOptions {
 }
 
-func (a *adminConfig) HostFactory(host, kubeconfigPath string) cmdutil.Factory {
+func (a *adminConfig) FedClientSet(context, kubeconfigPath string) (*fedclient.Clientset, error) {
```
Call this function `FederationClientset`. Expand fed and lower s in set. It is better for readability and Clientset follows the convention elsewhere.

irfanurrehman replied:
updated

madhusudancs reviewed (commented):
A few minor comments.

Also, github won't let me reply to a comment on an updated hunk, so replying here:

> I have a slight confusion about this. Please help me understand better.
> Actually when I implemented it first, I did something similar what are asking for. I did extend the AdminConfig interface to add another method getconfig, which built the config on similar lines as is currently done to get hostfactory, and then got a separate client for fed control plane from that config. I however removed that in favor of simplicity of the current one.
> Also on futher checking, the HostFactory() impl gets the new factory from package kubectl/cmd which I guess "should" work with either federation API server or kube API server, for that particular version alike.
>
> I would any ways keep the impl as suggested by you also ready, and update this commit or not, based on your suggestion.

I agree about the simplicity. But the trouble here is, the client and the API types returned by `kubetcl/cmd.Factory` are for Kubernetes. It all works because of the way things are hooked up today. If those assumptions are violated in the future, they might break. There is no promise about stability here. `cmd.Factory` is an internal thing. It can change. So I think we shouldn't risk that.

irfanurrehman commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -229,7 +230,20 @@ func initFederation(cmdOut io.Writer, config util.AdminConfig, cmd *cobra.Comman
 		return err
 	}
 
+	fedPods := []string{}
```
done

irfanurrehman commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -576,6 +590,47 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
 }
 
+func waitForPods(clientset *client.Clientset, fedPods []string, namespace string) error {
+	for _, fedPod := range fedPods {
```
updated

irfanurrehman commented:
@madhusudancs updated as per your suggestions and squashed the review fix commits also. PTAL.

madhusudancs commented on federation/pkg/kubefed/init/init.go:
```diff
@@ -576,6 +586,43 @@ func createControllerManager(clientset *client.Clientset, namespace, name, svcNa
+		for _, pod := range podList.Items {
+			for _, fedPod := range fedPods {
+				if strings.HasPrefix(pod.Name, fedPod) && pod.Status.Phase == "Running" {
+					return true, nil
```
As I was pointing out in my previous comment, you will need additional logic here. What is happening here is, the poll loop exits due to the return statement as soon as a single `Running` pod is found. However, you need to verify that both the pods are running.

madhusudancs reviewed (commented):
@irfanurrehman one minor comment, everything else looks good. Please fix that and let's get this merged.

irfanurrehman commented:
@madhusudancs updated. Also apologies for taking your additional time.

madhusudancs commented:
@irfanurrehman no worries. Thanks for making these changes and for being patient through these reviews. LGTM!

madhusudancs commented:
@irfanurrehman bazel and verification tests are failing. Please take a look.

irfanurrehman commented:
@madhusudancs updated bazel build files. All tests seem to be passing now. Need lgtm label.

madhusudancs commented:
@irfanurrehman cool! Added the label.
